src/isomorphism/vf2.rs

```diff
@@ -41,6 +38,86 @@ use crate::iterators::NodeMap;
 
 type StablePyGraph<Ty> = StableGraph<PyObject, PyObject, Ty>;
 
+/// Returns `true` if we can map every element of `xs` to a unique
+/// element of `ys` while using `matcher` func to compare two elements.
+fn is_subset<T: Copy, F>(xs: &[T], ys: &[T], matcher: F) -> PyResult<bool>
+where
+    F: Fn(T, T) -> PyResult<bool>,
+{
+    let mut valid = vec![true; ys.len()];
+    for &a in xs {
+        let mut found = false;
+        for (&b, free) in ys.iter().zip(valid.iter_mut()) {
+            if *free && matcher(a, b)? {
+                found = true;
+                *free = false;
+                break;
+            }
+        }
+
+        if !found {
+            return Ok(false);
+        }
+    }
+
+    Ok(true)
+}
+
+#[inline]
+fn sorted<N: std::cmp::PartialOrd>(x: &mut (N, N)) {
+    let (a, b) = x;
+    if b < a {
+        std::mem::swap(a, b)
+    }
+}
+
+/// Returns the adjacency matrix of a graph as a dictionary
+/// with `(i, j)` entry equal to number of edges from node `i` to node `j`.
+fn adjacency_matrix<Ty: EdgeType>(
+    graph: &StablePyGraph<Ty>,
+) -> HashMap<(NodeIndex, NodeIndex), usize> {
+    let mut matrix = HashMap::with_capacity(graph.edge_count());
+    for edge in graph.edge_references() {
+        let mut item = (edge.source(), edge.target());
+        if !graph.is_directed() {
+            sorted(&mut item);
+        }
+        let entry = matrix.entry(item).or_insert(0);
+        *entry += 1;
+    }
+    matrix
+}
+
+/// Returns the number of edges from node `a` to node `b`.
+fn edge_multiplicity<Ty: EdgeType>(
+    graph: &StablePyGraph<Ty>,
+    matrix: &HashMap<(NodeIndex, NodeIndex), usize>,
+    a: NodeIndex,
+    b: NodeIndex,
+) -> usize {
+    let mut item = (a, b);
+    if !graph.is_directed() {
+        sorted(&mut item);
+    }
+    *matrix.get(&item).unwrap_or(&0)
+}
+
+/// Nodes `a`, `b` are adjacent if the number of edges
+/// from node `a` to node `b` is greater than `val`.
+fn is_adjacent<Ty: EdgeType>(
+    graph: &StablePyGraph<Ty>,
+    matrix: &HashMap<(NodeIndex, NodeIndex), usize>,
+    a: NodeIndex,
+    b: NodeIndex,
+    val: usize,
+) -> bool {
+    let mut item = (a, b);
+    if !graph.is_directed() {
+        sorted(&mut item);
+    }
+    *matrix.get(&item).unwrap_or(&0) >= val
```

Comment:
```suggestion
    edge_multiplicity(&graph, &matrix, a, b) >= val
```

I think the shorter way makes it more obvious